### ISSUE TITLE
【求新增】智能文档分析模块（MindSpore Viewer）

### ISSUE THREAD

**Issue by jiahaochen666** [mindspore-lab/mindocr#734]
# 简介
## 目的
MindSpore OCR 目前主要应用在银行、交通、金融等行业中，这些行业的业务中存在大量的版面分析、表格分析、图像中关键信息提取等可通过人工智能解决的任务。MindSpore OCR 中集成智能文档分析模块，可以极大的解决客户的需求，并提高 MindSpore OCR 套件本身的竞争力。
#### 1. 软件名称
智能文档分析模块 MindSpore Viewer
#### 2.	软件功能
版面分析、表格识别、关键信息抽取、版面复原
# 第零层设计描述
## 软件系统上下文定义
安装的环境要求：同 MindSpore OCR 安装的环境要求相同。
![image](https://github.com/user-attachments/assets/fa7deaf5-87d9-463f-a490-2cce6b2fe480)
如图1所示，原 MindSpore OCR 仓下主要的代码模块在 mindocr 目录下，需要添加 mindviewer 目录存放 MindSpore Viewer 主体代码，而 MindSpore Viewer 作为 MindSpore OCR 的一个子模块存在，在安装完 MindSpore OCR 后，用户可以通过 `from mindocr import mindviewer` 模块导入使用。**新增子模块对原mindocr的api无任何影响。**
 
#### 1. 遵循标准
遵循Apache 2.0
#### 2.	硬件限制
Ascend：910A、910B
GPU：支持cuda10.1以上
CPU
#### 3. 技术限制
GPU、CPU暂不支持MindSpore 2.3
#### 4. 系统配置
openEuler: >= 20.3
CentOS：>= 7.6
Ubuntu：>= 18.04
Windows 10
MacOS：10.15/11.3
## 技术方向演进
当前版本的MindSpore Viewer不会实现如下功能，但是可作为未来演进方向的一些参考，通过对比友商的OCR套件，我们得出如下几种优秀的OCR套件特性：
+ 可视化交互OCR：通过拖拽方式导入数据，并进行在线的文本、图像编辑
+ 操作系统层面OCR：如Windows 11推出的OCR可以识别Android/IOS投屏图像中的内容
+ 多语言OCR：当前只支持中文和英文，后续可增加其他语言的OCR模型，并实现语种之间的翻译
+ 与大模型结合OCR：如PP-ChatOCR可给予LLMs提取图像中的关键信息
# 第一层设计描述
## 总体结构
MindSpore Viewer的整体架构如图2所示
![image](https://github.com/user-attachments/assets/fde3a8fb-9d0c-4b03-ace3-b1d9d5234976)
#### 1. 模块分解
+ Image（图像处理）
负责输入图像的预处理以及处理过程中对图像的超分和锐化
+ Layout（版面分析）
负责将图像根据版面内容划分成图片、文字、标题、和表格四种区域，再进行下一步处理
+ Converter（格式转化）
负责将识别出来的pfd文档转换成word文档格式或相反
+ Recovery（版面恢复）
负责将图像、pdf内容像原文档一样输出到word文档中
+ KIE（关键信息抽取）
负责从文本或者图像中，抽取出关键信息作为OCR下游任务，如表单识别等
+ Table（表格分析）
负责从图像中识别出表格并识别表格中的文字以及排列方式
#### 2. 并发进程处理分解
MindSpore Viewer支持多进程并发处理，当前V1版本先支持单卡的训练及推理，V2版本开始加入多卡训练和推理。多进程特性在各模块后续的设计中加入，当前设计文档按单进程处理。
#### 1.	运行设计
![image](https://github.com/user-attachments/assets/fb5b34d6-2917-4fb7-a16f-7b3b26760993)
MindSpore Viewer运行流程图如图3所示，首先原始图像输入会经过Image模块做预处理（缩放、旋转、格式转化等），处理好后会根据任务分类进入不同的管道：
+ 版面分析任务：Layout模块将图像分割成表格、图片、文字区域，表格会送进Table模块进行结构化识别，图像会送进Image模块进行超分，文字区域可通过开关判断是否送进Image模块进行锐化，再送进MindOCR进行识别，然后通过Recovery模块将前三个区域重新整合，可通过开关决定是否进行格式转换，最后输出或保存处理完的文档。
+ 关键信息抽取任务：预处理好的图像会直接送进MindSpore OCR中进行文字识别，然后通过语义识别模块获取语义实体，最后经过关系抽取模块获取语义实体间的对应关系，从而提取需要的关键信息。
 
#### 2.	数据依赖设计
如图3所示，若输入为n x m的图像Numpy Array，通过缩放可能a x b大小，通过Layout分割后输出三个list，分别为图像中各种分类的具体顶点坐标及标签。通过MindOCR模块识别得出关键信息的词分类和内容。
# 第二层设计描述
#### 1.	总体结构
![image](https://github.com/user-attachments/assets/a0b0b696-6920-41f6-8e1a-99125ec1ab7b)
Image模块整体功能如图4所示，可对输入的图像进行预处理（旋转、缩放、格式转换）以及后处理（超分、锐化）。
#### 2.	模块分解描述
Rotate：负责图像位置矫正
Scale：负责图像大小缩放
Convert：负责图像格式转换
Super Resolution：负责图像清晰度提升
Sharpen：负责图像锐化
#### 3.	数据实体描述
![image](https://github.com/user-attachments/assets/54a23a29-b9a6-4f1a-931a-e919081fd9f9)
Image类有如下公共变量：
+ `data`：存储图像数据的变量，由cv2.imread接口获取数据
+ `resolution`：图像的分辨率，tuple类
+ `angle`：文字图像的旋转角度，可由用户自行输入，或者用MobileNetV3模型检测
+ `is_sharpen`：判断是否需要锐化，可由用户选择开关
+ `format`：图像格式，jpeg、png等

#### 4.	依赖性描述
外部第三方依赖库：opencv、numpy
#### 5.	接口描述
Image类有如下公共函数：
+ `rotate(angle)`：旋转图像，输入为旋转角度，可由用户输入，或者通过MobileNetV3模型检测是否旋转矫正
+ `scale(resolution)`：缩放图像到指定大小，resolution为输入的tuple(wide, height)
+ `convert(string)`：转换图像格式，string为指定的图像格式，当前主要实现jpeg和png的互转
+ `superResolution(float)`：超分模块，输入float为超分的倍数，可根据硬件性能采取普通插值法进行超分，或使用SRCNN模型进行超分（后续根据效果再进行一次模型的选型）。
+ `sharpen()`：文字图像的锐化，通过高斯模糊区分出图像的低频和高频区，将高频信号与原始图像叠加达到图像增强的效果。
#### 1.	总体结构
![image](https://github.com/user-attachments/assets/0d62bfd0-88ea-4eb6-a933-3ecaaa2adf0f)
Layout模块整体功能如图5所示，预处理后的图像经过Layout模块做版面分析后切割成Table、Figure、Text区域。
#### 2.	模块分解描述
Layout模块无需分解额外子模块，所有功能包括数据前后处理、模型加载以及推理均在模块内处理完成。
#### 3.	数据实体描述
![image](https://github.com/user-attachments/assets/d0e0331f-6988-4319-a85a-3024b48fd973)
Layout模块下新建LayoutAnalyzer类，该类有如下几个初始化变量：
+ `model_path`：string类型，为模型的ckpt文件的路径，用于加载模型
+ `input_tensor`：Tensor类，用于图像数据的输入
+ `output_tensor`：Tensor类，用于模型推理后的结果输出
+ `preprocess_op`：Preprocessor类，用于图像数据前处理，需要从infer/text中import，需要在该类中新增对YOLOv8模型数据的前处理
+ `postprocess_op`：Postprocessor类，用于推理结果的处理，需要从infer/text中import，需要在该类中新增对YOLOv8模型推理的后处理
#### 4.	依赖性描述
Layout模块需要依赖cv2、mindocr中py_infer的LayoutPredictor、依赖infer/text下的preprocessor和postprocessor模块
#### 5.	接口描述
+ `__init__(self)`：类的构造函数，初始化上述几个变量
+ `__call__(self, img)`：类的调用函数，该类的object可以像函数一样被调用，可接受图像数据为入参，输出为模型推理的结果，输出类型为List，List中的每一项为分割后图像的类型、顶点数据、以及概率。
#### 1.	总体结构
![image](https://github.com/user-attachments/assets/8c7d5e2d-c88d-469f-9b4f-65cb235479f6)
如上图所示，Converter模块可以将pdf格式或docx格式的文件转化成docx或pdf格式的文件。
#### 2.	模块分解描述
+ Parallel：多进程模块，该模块加速处理文件格式的转换
+ pdf2docx：pdf转docx模块
+ docx2pdf：docx转pdf模块
#### 3.	数据实体描述
![image](https://github.com/user-attachments/assets/ea3fd5ab-d3f0-44b0-bbcc-d519fe34b061)
创建Parallel类，该类初始化两个变量：
+ `pdf2docx`：Pdf2docx类，该类初始化时一个变量为pdf文件路径，一个为pdf2docx的api
+ `docx2pdf`：Docx2pdf类，该类初始化是一个变量为docx文件路径，一个为docx2pdf的api
+ `dir`：文件夹路径，将文件夹下的所有文件做转换
+ `worker`：进程数量，默认为1（1的时候不开启多进程）
#### 4.	依赖性描述
依赖第三方库pdf2docx、docx2pdf
#### 5.	接口描述
+ `run()`：Parallel模块的外部接口，同时作为Converter模块的主要调用接口，会开启多进程任务，加载文件并做格式转换。
#### 1.	总体结构
![image](https://github.com/user-attachments/assets/286f9398-ad89-406b-be71-220e50102cc3)
Recovery(版面恢复)模块整体功能如图所示，将输入的图片、pdf内容经过版面分析、表格识别等技术，将处理后的结果，依照原文档位置信息重建到 word 文档中。
 
#### 2.	模块分解描述
Recovery模块无需分解额外子模块。
#### 3.	数据实体描述
![image](https://github.com/user-attachments/assets/49e62a93-69d7-40cd-aff7-f447b748d43d)
Recovery模块下新建MindViewerResult、BBox、ProcessedResult、ImageProcessedResult、HTMLProcessedResult、TextProcessedResult类：
MindViewerResult包含以下变量：
+ `bbox`: BBox类型，表示一个边界框(Bounding Box),用于定义页面中某个元素的位置和大小
+ `processed_result`: ProcessedResult类型，表示对不同类别(如文本、图像、表格)处理后的结果

BBox包含以下变量：
+ `x_min`: float类型，表示左上角的X坐标
+ `y_min`: float类型，表示左上角的Y坐标
+ `x_max`: float类型，表示右下角的X坐标
+ `y_max`: float类型，表示右下角的Y坐标

ProcessedResult是一个通用实体(父类)，包含以下变量：
+ `category_type`: string类型，表示处理结果的类型

ImageProcessedResult、TableProcessedResult、TextProcessedResult分别表示不同类型的处理结果，并继承自ProcessedResult。每个具体的子类包含特定类型内容：
+ `content`: string类型，表示处理的具体内容
#### 4.	依赖性描述
外部第三方依赖库：docx、bs4、html
#### 5.	接口描述
![image](https://github.com/user-attachments/assets/b90daf22-26a6-4059-9ef0-bae9f5f56417)
Recovery模块下新建LayoutRecovery类，该类负责对MindViewerResult实例进行初始化、排序，并输出到文档中，下面是该类的函数方法：
+ `__init__(self, elements)`:
	+ 功能:类的构造函数，用于初始化LayoutRecovery对象
	+ 输入: elements,MindViewerResult实例的列表
	+ 输出: 无。
+ `sort_by_layout(self)`:
	+ 功能: 根据版面信息(bbox)对MindViewerResult实例进行排序。
	+ 输入: 无。
	+ 输出: 排序后的MindViewerResult实例列表。
+ `export_to_doc(self, output_path)`:
	+ 功能: 将排序后的MindViewerResult实例输出到doc文档中。
	+ 输入: output_path，输出文档的路径。
	+ 输出: 无。
+ `_export_to_doc(self, elements, output_path)`:
	+ 功能: 辅助方法，将MindViewerResult实例输出到doc文档。
	+ 输入:elements，排序后的MindViewerResult实例列表	output_path: 输出文档的路径。
	+ 输出: 无。
#### 1.	总体结构
KIE(Key Information Extraction,关键信息抽取) 模块整体功能如图所示，指在图像中，抽取出关键信息。
![image](https://github.com/user-attachments/assets/13fbff40-f376-4243-9479-b486985c595c)
#### 2.	模块分解描述
KIE包含关系抽取(Relation Extraction, RE)和结构化实体识别(Structured Entity Recognition, SER)两个子模块：
+ 关系抽取(RE): 识别文本中的实体，并推断这些实体之间的关系。
+ 结构化实体识别(SER): 从非结构化文本中识别并提取结构化的实体(如人名、地名、组织名等)。
#### 3.	数据实体描述
![image](https://github.com/user-attachments/assets/54ea97e9-9c21-4493-8898-78e56e503144)
KIE模块中SER模块可使用已训练好的 VI_LAYOUTXLM 及 LAYOUTXLM 模型，ckpt可以直接从MindSpore OCR仓提供的链接下载并加载，接受一组待推理的图片列表。
 
KIE模块下新建KIEOutput、SEROutput、REOutput类，其中 SEROutput、REOutput 分别表示SER模块及RE模块的处理结果，并继承自 KIEOutput ：
KIEOutput 包含以下变量：
+ `input_ids`: List[int]类型，序列化数据的编码 ID 列表
+ `bbox`: BBox类型，表示一个边界框(Bounding Box),用于定义页面中某个元素的位置和大小
+ `attention_mask`: List[int]类型，注意力掩码
+ `token_type_ids`: List[int]类型，区分输入文本的不同段落或句子
+ `image`:Image 输入的图像数据

SEROutput子类继承KIEOutput 所有类型，并包含以下特定属性：
+ `labels`: List[str]类型，识别出的实体标签

REOutput子类继承KIEOutput 所有类型，并包含以下特定属性：
+ `question`: List[str]类型，表示关系抽取任务中涉及的问题
+ `question_label`: List[str]类型，question对应的标签
+ `answer`: List[str]类型，表示关系抽取任务中提取的答案
+ `answer_label`: List[str]类型，answer对应的标签
+ `relation_label`:List[str]类型，两个实体之间的关系标签，定义实体对之间的关系。
#### 4.	依赖性描述
SER模块依赖cv2、mindocr中infer/text下的predict_ser模块
RE模块在mindocr 中暂无推理工具，需要对infer/text下的工具进行补全，如predict_re。
#### 5.	接口描述
![image](https://github.com/user-attachments/assets/0d661acb-f9f6-4029-93ec-c5782748a7ff)
KIE模块下新建SerPredictor类，该类有如下几个初始化变量：
+ `img_path`: string类型，为输入待预测图片的路径
+ `model_path`：string类型，为模型的ckpt文件的路径，用于加载模型
+ `algorithm`：string类，使用到的算法
+ `output_path`: string类，预测图片输出的路径

同时创建两个类函数：
+ `__init__(self)`：类的构造函数，初始化上述几个变量
+ `__call__(self, img_path, output_path)`：类的调用函数，该类的object可以像函数一样被调用，可接受图像路径为入参，将结果输出到具体路径。
#### 1.	总体结构
Table模块整体功能如图所示，通过表格结构识别模块（Structure Recognition），解析出表格的行列结果，并通过表格内容识别模块（Content Recognition）识别和提取表格中的信息，最后支持将表格导出为html、excel、doc 或者其他格式。
![image](https://github.com/user-attachments/assets/f325af84-f994-4235-ab45-c7c39519d984)
#### 2.	模块分解描述
Table模块分为两个子模块：
Structure Recognition: 表格结构识别模块
Content Recogintion: 表格内容识别模块
#### 3.	数据实体描述
![image](https://github.com/user-attachments/assets/93309822-f2a7-48d6-baf7-ebf1ca172b0b)
Table模块下新建TableOutput、StructureOutput、ContentOutput类，其中 StructureOutput 、 ContentOutput 分别表示SER模块及RE模块的处理结果，并继承自 TableOutput ：
TableOutput 包含以下变量：
+ `bbox`: BBox类型，表示一个边界框(Bounding Box),用于定义页面中某个元素的位置和大小
+ `image`:Image 输入的图像数据

StructureOutput 子类继承 TableOutput 所有类型，并包含以下特定属性：
+ `shape`: 表格大小
+ `structure`：表格行列结构

ContentOutput 子类继承 TableOutput 所有类型，并包含以下特定属性：
+ `content`: 表格内容
#### 4.	依赖性描述
Table模块在 mindocr 中暂无推理工具，需要对infer/text下的工具进行补全，如predict_table。
#### 5.	接口描述
![image](https://github.com/user-attachments/assets/d701dc0f-8b4f-4feb-ac0e-6f1393a3e2a9)
Table模块下新建TableAnalyzer类,包含以下接口：
+ `Structure_recognition`: 输入图像，根据模型推理，返回表格的行列结构。
+ `Content_recognition`: 输入图像，根据模型推理，输出表格的单元格内容。
+ `Reconsturct_table`: 接受表格结构和内容信息，生成指定格式的结构化表格输出。
# 配置和控制
## 启动
新建start.py文件作为主函数入口，用户可通过以下两个参数创建不同的任务流水线：
+ --predict：版面分析任务
+ --kie：关键信息提取任务

例如如下命令行可启动版面分析任务：
```bash
python start.py --predict image.png
```
## 关闭
新建clean.py文件，当任务处理完或者通过`Ctrl-C`中断时，可调用该文件清理环境。
# 错误处理
## 系统错误
若因为系统错误导致程序中断（如显存不足），可指导用户调用clean.py清理环境释放内存。
## 接口错误
新增error.log文件，如果MindSpore Viewer代码出现错误，抛出错误并输出错误原因到error.log日志中。
